bees/mumblebee/mumblebee.go

```diff
@@ -0,0 +1,240 @@
+ *    GNU Affero General Public License for more details.
+ *
+ *    You should have received a copy of the GNU Affero General Public License
+ *    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ *    Authors:
+ *      Martin Schlierf <go@koma666.de>
+ */
+
+// Package mumblebee is a Bee that can connect to a Mumble/XMPP server.
+package mumblebee
+
+import (
+	"crypto/tls"
+	"net"
+	"strconv"
+	"strings"
+
+	"github.com/muesli/beehive/bees"
+	"layeh.com/gumble/gumble"
+	"layeh.com/gumble/gumbleutil"
+)
+
+// MumbleBee is a Bee that can connect to a Mumble server.
+type MumbleBee struct {
+	bees.Bee
+
+	// channel signaling mumble connection status
+	connectedState chan bool
+
+	client *gumble.Client
+
+	server   string
+	user     string
+	password string
+	insecure bool
+	channel  string
+}
+
+// Action triggers the action passed to it.
+func (mod *MumbleBee) Action(action bees.Action) []bees.Placeholder {
+	outs := []bees.Placeholder{}
+
+	switch action.Name {
+	case "send_to_user":
+
+		user, text := "", ""
+
+		action.Options.Bind("user", &user)
+		action.Options.Bind("text", &text)
+
+		mod.client.Users.Find(user).Send(text)
+
+	case "send_to_channel":
+
+		channel, text := "", ""
+
+		action.Options.Bind("channel", &channel)
+		action.Options.Bind("text", &text)
+
+		if channel != "" {
+			mod.client.Channels.Find(channel).Send(text, false)
+		} else {
+			mod.client.Self.Channel.Send(text, false)
+		}
+	case "move":
+
+		channel := ""
+
+		action.Options.Bind("channel", &channel)
+		mod.move(channel)
+
+	default:
+		panic("Unknown action triggered in " + mod.Name() + ": " + action.Name)
+	}
+
+	return outs
+}
+
+func (mod *MumbleBee) move(channel string) {
+	channel = strings.TrimSpace(channel)
+	if len(channel) > 0 {
+		channels := strings.Split(channel, "/")
+		result := mod.client.Channels.Find(channels...)
+		if result != nil {
+			mod.client.Self.Move(result)
+			mod.Logln("Moved to channel:", channel)
+		} else {
+			mod.Logln("Moved to channel:", channel, " failed. (channel not found)")
+		}
+	}
+}
+
+// Run executes the Bee's event loop.
+func (mod *MumbleBee) Run(eventChan chan bees.Event) {
+	if len(mod.server) == 0 {
+		return
+	}
+
+	// channel signaling mumble connection status
+	mod.connectedState = make(chan bool)
+
+	host, port, err := net.SplitHostPort(mod.server)
+	if err != nil {
+		host = mod.server
+		port = strconv.Itoa(gumble.DefaultPort)
+	}
+
+	config := gumble.NewConfig()
+	config.Username = mod.user
+	config.Password = mod.password
+	address := net.JoinHostPort(host, port)
+
+	var tlsConfig tls.Config
+
+	if mod.insecure {
+		tlsConfig.InsecureSkipVerify = true
+	}
+
+	config.Attach(gumbleutil.Listener{
+		TextMessage: func(e *gumble.TextMessageEvent) {
+			ev := bees.Event{
+				Bee:  mod.Name(),
+				Name: "message",
+				Options: []bees.Placeholder{
+					{
+						Name:  "channel",
+						Type:  "[]string",
+						Value: e.Channels,
```

Comment:
I'm confused. ```e.Channels``` seems to indicate a message can be received on multiple channels? Isn't it always just one channel?